### DISCUSSION TITLE
How to cancel a async job after async engine return ASYNC_PAUSE?

### DISCUSSION THREAD

**Discussion by zh-jq-b** [openssl/openssl#23158] [Q&A]
If we have alloc some data **P**, and pass it to SSL_set_async_callback_arg / ASYNC_WAIT_CTX_set_callback,
then we wait sometime for the async to finish after ASYNC_PAUSE. But if it timeout, we need to cancel the async operation,
so what should we do to free **P**?
There is no API to tell the async engine to cancel the async job, so there will be a use-after-free issue if the async engine call the callback with that **P** as arg.

**Comment by t8m** [comment]
@mattcaswell any idea?

**Comment by zh-jq-b** [comment]
A denial of service may be caused by a client sending SSL messages slowly to the server, which has been configured with ASYNC mode enabled and have set a timeout to accept the full client handshake.
And in the cancel case, it seems the ASYNC_JOB is also not returned to the job pool?

**Comment by mattcaswell** [comment]
There is no API to cancel an ASYNC job at the current time. It would have to be implemented via some out-of-band mechanism to tell the engine to cancel a request such that the next time the cancelled job is resumed it just fails immediately (i.e. the ASYNC job is then effectively cleaned up in the normal manner and returned to the pool).

**Comment by nhorman** [comment]
Given that the ASYNC api is implemented via user context switching rather than a separate thread, can it be considered 'safe' to treat the cb arg as being referenced only by the function that got the ASYNC_PAUSE return code?  i.e. can the arg be set to NULL at that point and freed so that on the next job start the impacted thread context would see the NULL value and return immediately?

**Comment by mattcaswell** [comment]
> Given that the ASYNC api is implemented via user context switching rather than a separate thread, can it be considered 'safe' to treat the cb arg as being referenced only by the function that got the ASYNC_PAUSE return code? i.e. can the arg be set to NULL at that point and freed so that on the next job start the impacted thread context would see the NULL value and return immediately?

The cb arg is entirely opaque to the engine and could be NULL even in the normal course of events. It is not possible to change the arg after the job has started via libssl, so there is no way for the engine to know that it has been freed.

**Comment by nhorman** [comment]
Right, I get that, but I was more asking if there was a safe method to free the callback arg prior to the job completing.  The OP was curious to know if, after some timeout event, if the cbarg could be freed safely prior to job completion.

Nominally, you can do that by canceling the job first (which, as you've noted, you can't do here).  My question was more, is there a convenient way to free the callback arg prior to the async_cb getting called? i was thinking that you could try to rely on the fact that, because the async lib uses userspace contexts rather than threads, you could call ASYNC_pause_job to ensure no job was running, then update the pointer, but looking at the code that seems risky

Alternatively I suppose you could implement the async_cb function that gets set such that a global lock is taken, and arg could be set to a void **, so that you could updated the pointed to object.

**Comment by zh-jq-b** [comment]
QAT_Engine use ASYNC_JOB across threads, which may lead to a crash, see https://github.com/intel/QAT_Engine/issues/292.
It will be better for the ASYNC APIs to make ASYNC_JOB and WAIT_CTX to be thread safe by using locks/ref-count. And for the callback args, a clean func may be registered with the ASYNC_JOB/WAIT_CTX to free it after finish.
